Lombiq.Hosting.Tenants.FeaturesGuard/Handlers/FeaturesEventHandler.cs

```diff
@@ -90,9 +88,20 @@ public async Task EnableConditionallyEnabledFeaturesAsync(IFeatureInfo featureIn
         // If Shell Descriptor's Features already contains a feature that is found in conditionalFeatures, remove it
         // from the list. Handle multiple conditional features as well.
         var featuresToEnable = allFeatures.Where(feature =>
-            conditionalFeatureIds.Contains(feature.Id) && !shellDescriptor.Features.Contains(new ShellFeature(feature.Id)));
+            conditionalFeatureIds.Contains(feature.Id) &&
+            !shellDescriptor.Features.Any(shellFeature => conditionalFeatureIds.Contains(shellFeature.Id)));
+
+        foreach (var feature in featuresToEnable)
+        {
+            if (_previouslyAddedFeatures.Contains(feature))
+            {
+                return;
+            }
```

Comment:
What if `featuresToEnable` has other, new features as well?

Reply:
I don't quite get the point of you question now. Could you give more detail on what do you mean?
featuresToEnable can't change during 1 handler execution. It is just querying the features that should be enabled depending on the condition features. We needed the list field here to store features that the handler have already turned on because OC seems to be unable to get the enabled features during run time.
By this I mean that shellDescriptor.Features will not include the Features that has just been enabled by the Handler.
So lets say the handler #1 execution return 50 enabled features from the shellDescriptor.
Then it will enable OrchardCore.Twitter.
Then the handler runs again. Execution #2 the shellDescriptor will again return 50 as the count of enabled features eventhough Twitter was enabled and it should be 51.
Imo this is an OC bug of somekind. This is why I called this a workaround. This was causing the problem as the code could never actually check if the feature is turned on because the list of enabled features never changed. And it tried to enabled features twice which caused the invalid serial number for shell descriptor error.

Reply:
`featuresToEnable` can have more than one item, right? If indeed, then it's possible that e.g. item 1 will already be in `_previouslyAddedFeatures`, and item 2 won't. In this case, the method will return after item 1, but item 2 is something that should be enabled.

Reply:
For info the `...ing` handlers are first called, then the shell descriptor is updated (immediately committed internally), then the `...ed` handlers are called but in a deferred task so in its own scope (the handlers are scoped services).

I don't think that's a good pattern to call `.EnableFeaturesAsync()` in a `...ed` handler (called for each feature) because it triggers another shell descriptor update, a new deferred task in which the scoped `...ed` handlers are called again.

Would need to look at it more in details but it looks like a new dependency graph. For info a feature can be marked as `EnabledByDependencyOnly` (manifest attribute) meaning that it is auto enabled if at least a dependent feature is enabled, and kept enabled as long as one dependent feature is enabled.

But you may need to define it from configuration (not manifest files), not fully sure but as I remember we have filters/handlers to config feature infos and/or define custom dependency strategies.

Reply:
@jtkech Thank you so much JT for sharing this it really helps. 
As for the `EnabledByDependecyOnly` part it would be good but this feature serves the purpose if one enables a condition feature then the conditional feature should be auto enabled but this does not mean that one could not directly enable the conditional feature.

Reply:
@jtkech Do you have any idea why `EnableFeaturesAsync()` throws an `InvalidOperationException` error when:
![image](https://user-images.githubusercontent.com/92299130/235620580-2debc750-20e7-4fc1-89e3-29242c0d63b2.png)
So now I am really strictly trying to avoid feature enabling duplication.
Which works because I am trying to turn on 3 features:
Twitter, Facebook, Forms
Here are the condition and conditinals: 
![image](https://user-images.githubusercontent.com/92299130/235620898-856909e5-a2bb-4a60-a062-6441b6858dd6.png)

Upon enabling Lombiq.ChartJs the feature turns on OrchardCore.Twitter and OrchardCore.Facebook and it works perfectly.

Then comes Lombiq.UIKit.
The handler is able to extract the feature that have already been enabled (Twitter and Facebook) and it returns OrchardCore.Forms on to the `featuresToEnable` collection.
Then `await.EnableFeaturesAsync(featuresToEnable)` containing only `OrchardCore.Forms` is throwing an error.
So basically not only is `await EnableFeaturesAsync()` throwing an exception when you try to enable a feature that has already been enabled but even if the feature is not enabled.
I don't understand why it works for the first execution when the condition is `Lombiq.ChartJs` and why it breaks at `Lombiq.UIKit`

Reply:
P.S the whole above written story happens when I create a new Tenant from the Default tenant.
So this handler will run for every single feature that is enabled by default when creating a tenant.
It activates Lombiq.ChartJs first and after that some point it will activate Lombiq.UIKit.
When it activates Lombiq.UIKit the handler will fail with what I have described above

Reply:
This is the exception:

```
2023-05-06 14:47:16.9259|Test|0HMQE600KNIBJ||OrchardCore.Environment.Shell.ShellFeaturesManager|ERROR|IFeatureEventHandler thrown from Lombiq.Hosting.Tenants.FeaturesGuard.Handlers.FeaturesEventHandler by InvalidOperationException System.InvalidOperationException: Invalid serial number for shell descriptor
   at OrchardCore.Environment.Shell.Data.Descriptors.ShellDescriptorManager.UpdateShellDescriptorAsync(Int32 priorSerialNumber, IEnumerable`1 enabledFeatures)
   at OrchardCore.Environment.Shell.ShellDescriptorFeaturesManager.UpdateFeaturesAsync(ShellDescriptor shellDescriptor, IEnumerable`1 featuresToDisable, IEnumerable`1 featuresToEnable, Boolean force)
   at OrchardCore.Environment.Shell.ShellFeaturesManagerExtensions.EnableFeaturesAsync(IShellFeaturesManager shellFeaturesManager, IEnumerable`1 features, Boolean force)
   at Lombiq.Hosting.Tenants.FeaturesGuard.Handlers.FeaturesEventHandler.EnableConditionallyEnabledFeaturesAsync(IFeatureInfo featureInfo) in E:\ShortPath\OSOCE2\src\Modules\Lombiq.Hosting.Tenants\Lombiq.Hosting.Tenants.FeaturesGuard\Handlers\FeaturesEventHandler.cs:line 102
   at OrchardCore.Modules.InvokeExtensions.InvokeAsync[TEvents,T1](IEnumerable`1 events, Func`3 dispatch, T1 arg1, ILogger logger)    at OrchardCore.Environment.Shell.Data.Descriptors.ShellDescriptorManager.UpdateShellDescriptorAsync(Int32 priorSerialNumber, IEnumerable`1 enabledFeatures)
   at OrchardCore.Environment.Shell.ShellDescriptorFeaturesManager.UpdateFeaturesAsync(ShellDescriptor shellDescriptor, IEnumerable`1 featuresToDisable, IEnumerable`1 featuresToEnable, Boolean force)
   at OrchardCore.Environment.Shell.ShellFeaturesManagerExtensions.EnableFeaturesAsync(IShellFeaturesManager shellFeaturesManager, IEnumerable`1 features, Boolean force)
   at Lombiq.Hosting.Tenants.FeaturesGuard.Handlers.FeaturesEventHandler.EnableConditionallyEnabledFeaturesAsync(IFeatureInfo featureInfo) in E:\ShortPath\OSOCE2\src\Modules\Lombiq.Hosting.Tenants\Lombiq.Hosting.Tenants.FeaturesGuard\Handlers\FeaturesEventHandler.cs:line 102
   at OrchardCore.Modules.InvokeExtensions.InvokeAsync[TEvents,T1](IEnumerable`1 events, Func`3 dispatch, T1 arg1, ILogger logger)
```

As apparent from the stack trace, `FeaturesEventHandler` is called recursively, since the event is raised also after a feature is enabled from the event itself.

Try the `EnablingAsync` event instead, as JT above perhaps suggests.

Reply:
EnablingAsync I have tried and it wasn't even hitting the breakpoint for me. But I will double check although I am quite sure this was the case

Reply:
If it's not usable, try to figure out how you can prevent this recursive enable, because I guess that's causing the issue.

Reply:
At first I thought that the problem is due to the fact that if there are multiple conditions for 1 feature the Handler will try to enable the feature as many times and as the filtering for already enabled features is not working this was causing error.
I have implemented a filtering solution which , from debugging at least shows, prevents the handler to enable the same feature twice. After I have come up with this solution and I was testing it it turned out that even after filtering the `Invalid serial number for shell descriptor` issue is still present. I see from debugging that the collection passed to `EnableFeaturesAsnyc()` is clearly not containing the previous features just the one that is not yet enabled and the error still pops.

Reply:
Okay so now I have debugged using `EnablingAsync` and the reason I was not hitting any breakpoints because my breakpoints were placed in a position that is really not hit. I put some at the begging and it seems that `EnablingAsync` is going through a handful of features only like really core OC features. I manually watched all the executions and features that are enabled by the recipe were not included. So I never hit the condition features.

Reply:
Demo step by step:

Features I want to enable:
![image](https://user-images.githubusercontent.com/92299130/236632712-036efb0f-0b3a-42ef-bac5-ff8c0c071564.png)

Here Facebook and Twitter is being enabled and the feature the Handler is hitting is Lombiq.ChartJs
![image](https://user-images.githubusercontent.com/92299130/236632730-36c58599-e1c0-4eb3-997f-911b803b0874.png)

Now the feature is Lombiq.UIKit. You can see the collection only has OrchardCore.Forms and the other 2 has successfully been removed from the collection and is ready to be passed to `EnableFeaturesAsync()`
![image](https://user-images.githubusercontent.com/92299130/236632799-4b376794-e4e0-4aa3-ae5d-a8fe70a90fd4.png)

Eventhough it clearly got a unique collection with a feature that hasn't been enabled yet the error occures:
![image](https://user-images.githubusercontent.com/92299130/236632859-095a26f9-a7e1-419b-a4ce-10047c37f106.png)

Reply:
Try to understand the root cause of the exception by working back from there. You can use the debugger for that, see https://lombiq.atlassian.net/wiki/spaces/DEV/pages/1400143919/Debugging+a+NuGet-based+Orchard+Core+solution.

Reply:
@Psichorex Oops sorry I missed your last comments, I will try to look at it this night.

In the meantime if I understood correctly the logic, e.g. if `.UIKit` is enabled, `.Forms` should be enabled and can't be disabled, if `.UIKit` is disabled `.Forms` can be enabled or disabled, right?

For info, when we update the shell descriptor we use `GetOrCreateMutableAsync<>()` which returns a new descriptor instance (other updates will get this same instance) on which we increment the serial number. But when `ShellFeaturesManager` update again the shell descriptor it still use the serial number of its originally resolved `_shellDescriptor` (considered as immutable).

Here we could have also used `GetOrCreateMutableAsync<>()`, but when we introduced the usage of `IDocumentStore` with its own scoped caches we didn't change this behavior. I thought about it but make sense that you can't update the shell descriptor twice in the same shell scope.

Also in a `...ed` handler executed in a deferred task in its own scope, the shell descriptor has been already saved, we do this to only call the handler in that case.

Reply:
Okay, I looked at it, I suggest to use an `IShellDescriptorManagerEventHandler` that updates the shell descriptor, one problem is that `ShellDescriptorManager` does a `.CommitAsync()` before caliing the handlers, this is done because `ShellHost` (being currently the only one handler) releases the shell so that a subsequent request will build a new shell. So the `ShellDescriptor` needs to be committed before, otherwise `ShellHost` may register a shell that was built with a wrong shell descriptor.

So when we update the `ShellDescriptor` we need to commit again and release the shell again at the very end. Then we would need to call the features enable/disable handlers and so on, repeating what the existing services do.

So I opted to do it in a separated deferred task using its own scope, so as it would be done by another request but here initiated by the feature guarding, and then by using the existing services

Here a testing code where I only check 2 features but just to show the idea.

    .AddOrchardCms()
    .ConfigureServices(services =>
        services.AddScoped<IShellDescriptorManagerEventHandler, ShellFeaturesGuard>())


    public class ShellFeaturesGuard : IShellDescriptorManagerEventHandler
    {
        public Task ChangedAsync(ShellDescriptor descriptor, ShellSettings settings)
        {
            var conditions = new Dictionary<string, string[]>
            {
                ["OrchardCore.Feeds"] = new[] { "OrchardCore.Apis.GraphQL" }
            };

            var featureIdsToEnable = new HashSet<string>();
            foreach (var kvp in conditions.ToArray())
            {
                if (descriptor.Features.FirstOrDefault(f => f.Id == kvp.Key) is not null)
                {
                    continue;
                }

                foreach (var condition in kvp.Value)
                {
                    if (descriptor.Features.FirstOrDefault(f => f.Id == condition) is null)
                    {
                        continue;
                    }

                    featureIdsToEnable.Add(kvp.Key);

                    break;
                }
            }

            if (featureIdsToEnable.Count > 0)
            {
                ShellScope.AddDeferredTask(async scope =>
                {
                    var shellFeaturesManager =
                        scope.ServiceProvider.GetRequiredService<IShellFeaturesManager>();

                    var featuresToEnable = (await shellFeaturesManager
                        .GetAvailableFeaturesAsync())
                        .Where(feature => featureIdsToEnable.Contains(feature.Id))
                        .ToArray();

                    if (featuresToEnable.Length > 0)
                    {
                        await shellFeaturesManager.EnableFeaturesAsync(featuresToEnable, true);
                    }
                });
            }

            return Task.CompletedTask;
        }
    }

Reply:
@jtkech Thank you JT! It seems that using `AddDeferredTask` did make the difference.

Reply:
@Psichorex Okay good to know!

Yes, I was suggesting an `IShellDescriptorManagerEventHandler` but if it works by still using an `IFeatureEventHandler` that's okay. Take care about recursions that are still possible, for example if you need to both enable and disable the same feature because of a bad options configuration.

---

Hmm one drawback I see is that if we enable multiple features in one pass, we may trigger multiple deferred tasks. So one suggestion would be that in your scoped handler each time a Condition is enabled/disabled you trigger the deferred task but only once by using a local boolean.

Then do all the logic in the deferred task, first resolve the `ShellDescriptor`, yes better anyway to retrieve it from the deferred task to be sure of the current state, can be injected directly by using.

    scope.ServiceProvider.GetRequiredService<ShellDescriptor>()

Then lookup on your options dictionary, if a Conditional is or is not in the descriptor check if a Condition is enabled or not, and update features to enable/disable hashsets. Maybe also check that a feature is not in both collections, otherwise remove from one + maybe a log message.

Then update the features by using the resolved feature manager. The above could be done by a service injecting the descriptor and all it needs so that you only have to resolve this service, and it could also be the `IFeatureEventHandler` (not same scoped instance) that triggers the deferred task.

---

Looking at the code, it seems that when we disable all Conditions we also want to disable the Conditional, if so it's really like if the Conditions has a dependency on the Conditional and if the Conditional is EnabledByDependencyOnly. So just for info I tried the following that works.

    ocBuilder.ApplicationServices
        .AddSingleton<IFeatureBuilderEvents, FeatureGuardBuilderEvents>()


    public class FeatureGuardBuilderEvents : FeatureBuilderEvents
    {
        public override void Building(FeatureBuildingContext context)
        {
            if (context.FeatureId == "OrchardCore.Apis.GraphQL")
            {
                var dependencies = context.FeatureDependencyIds.ToHashSet();
                if (dependencies.Add("OrchardCore.Feeds"))
                {
                    context.FeatureDependencyIds = dependencies.ToArray();
                }
            }

            if (context.FeatureId == "OrchardCore.Feeds")
            {
                context.EnabledByDependencyOnly = true;
            }
        }
    }

The only concern is that it may introduce a circular feature dependency if the Conditional also depends on the Condition, for example if `Media` should enable `Azure.Media` that depends on `Media.Cache` that also depends on `Media`. It would not fail but the order of services registrations (as drivers / handlers) and then the order of executions may change.